File: implement-shell-tools/ls/ls.py
import argparse
import os
import sys

parser = argparse.ArgumentParser(prog="ls", description="List directory contents")
parser.add_argument("-1", dest="one_per_line", action="store_true",
                    help="List one file per line")
parser.add_argument("-a", "--all", action="store_true",
                    help="Do not ignore entries starting with .")
parser.add_argument("paths", nargs="*", default=["."],
                    help="The files or directories to list")

args = parser.parse_args()


def entries(directory):
    names = os.listdir(directory)
    if args.all:
        names = names + [".", ".."]
    else:
        names = [name for name in names if not name.startswith(".")]
    return sorted(names)


files = []
directories = []
for path in args.paths:
    if os.path.isdir(path):
        directories.append(path)
    elif os.path.exists(path):
        files.append(path)
    else:
        print(f"ls: cannot access '{path}': No such file or directory",
              file=sys.stderr)

files.sort()
directories.sort()

show_headers = len(args.paths) > 1
printed_anything = False

for path in files:
    print(path)
    printed_anything = True

for directory in directories:
    if show_headers:
        if printed_anything:
            print()
        print(f"{directory}:")
    for name in entries(directory):
        print(name)
    printed_anything = True

File: implement-shell-tools/ls/ls.mjs
import process from "node:process";
import { promises as fs } from "node:fs";
import { parseArgs } from "node:util";

const { values, positionals } = parseArgs({
  options: {
    one: { type: "boolean", short: "1", default: false },
    all: { type: "boolean", short: "a", default: false },
  },
  allowPositionals: true,
});

const showAll = values.all;
const paths = positionals.length > 0 ? positionals : ["."];

function stripPunctuation(name) {
  return name.replace(/[^\p{L}\p{N}]/gu, "");
}

function compareNames(a, b) {
  const result = stripPunctuation(a).localeCompare(stripPunctuation(b));
  return result !== 0 ? result : a.localeCompare(b);
}

const files = [];
const directories = [];

for (const path of paths) {
  try {
    const stats = await fs.stat(path);
    if (stats.isDirectory()) {
      directories.push(path);
    } else {
      files.push(path);
    }
  } catch {
    console.error(`ls: cannot access '${path}': No such file or directory`);
    process.exitCode = 2;
  }
}

files.sort(compareNames);
directories.sort(compareNames);

const needHeaders = directories.length > 1 || files.length > 0;
let printedSomething = false;

for (const file of files) {
  console.log(file);
  printedSomething = true;
}

for (const dir of directories) {
  let entries = await fs.readdir(dir);

  if (showAll) {
    entries = [".", "..", ...entries];
  } else {
    entries = entries.filter((entry) => !entry.startsWith("."));
  }
  entries.sort(compareNames);

  if (needHeaders) {
    if (printedSomething) {
      console.log("");
    }
    console.log(`${dir}:`);
  }

  for (const entry of entries) {
    console.log(entry);
  }
  printedSomething = true;
}
